### PR TITLE
Fix audit feedback on committee creation

### PR DESCRIPTION
**Pull request by dariorussi**
## Description 

As reported by audit if a validator registers with an already used pubkey for the bridge (whether maliciously or not) the creation of the committee will fail and as such an end of epoch would fail attempting to create the committee.
Moreover there is no way at the moment to flush the proposed committee (in that respect, should we have an API for it?) and it would require an emergency release to take care of the problem.
This should prevent the problem from happeing.

## Test Plan 

Added unit test

---
If your changes are not user-facing and do not break anything, you can skip the following section. Otherwise, please briefly describe what has changed under the Release Notes section.

### Type of Change (Check all that apply)

- [ ] protocol change
- [ ] user-visible impact
- [ ] breaking change for a client SDKs
- [ ] breaking change for FNs (FN binary must upgrade)
- [ ] breaking change for validators or node operators (must upgrade binaries)
- [ ] breaking change for on-chain data layout
- [ ] necessitate either a data wipe or data migration

### CODE REVIEW

File: crates/sui-framework/packages/bridge/sources/committee.move [hunk 1]
```diff
@@ -164,9 +165,35 @@ module bridge::committee {
             registration
         };
 
+        // check uniqeuness of the bridge pubkey.
+        // `try_create_next_committee` will abort if bridge_pubkey_bytes are not unique and
+        // that will fail the end of epoch transaction (possibly "forever", well, we
+        // need to deploy proper validator changes to stop end of epoch from failing).
+        assert!(check_uniqueness_bridge_keys(self, bridge_pubkey_bytes), EDuplicatePubkey);
```
Comment (dariorussi) [inline crates/sui-framework/packages/bridge/sources/committee.move:172]: I could just call the API here and have the `assert` inside `check_uniqueness_bridge_keys` 
if anybody feels strongly about it let me know

File: crates/sui-framework/packages/bridge/sources/committee.move [hunk 2]
```diff
@@ -406,6 +433,32 @@ module bridge::committee {
...
+
+        let mut system_state = test_scenario::take_shared<SuiSystemState>(&scenario);
+
+        // validator registration
+        register(&mut committee, &mut system_state, hex::decode(VALIDATOR1_PUBKEY), b"", &tx(@0xA, 0));
+        register(&mut committee, &mut system_state, hex::decode(VALIDATOR1_PUBKEY), b"", &tx(@0xC, 0));
+
+        test_utils::destroy(committee);
+        test_scenario::return_shared(system_state);
+        test_scenario::end(scenario);
```
Comment (manolisliolios) [inline crates/sui-framework/packages/bridge/sources/committee.move:459]: You could avoid non-needed code on failure-expecting tests by aborting (`abort 1337`)

File: crates/sui-framework/packages/bridge/sources/committee.move [hunk 3]
```diff
@@ -164,9 +165,35 @@ module bridge::committee {
...
+        let mut count = self.member_registrations.size();
+        let mut found = false;
+        while (count > 0) {
+            count = count - 1;
+            let (_, registration) = self.member_registrations.get_entry_by_idx(count);
+            if (registration.bridge_pubkey_bytes == bridge_pubkey_bytes) {
+                if (found) {
+                    return false
+                };
+                found = true;
```
Comment (manolisliolios) [inline crates/sui-framework/packages/bridge/sources/committee.move:191]: nit: I'd probably pass down the sender's address too and check as the "found" thing messed with my brain (but could be me malfunctioning)

`if (registration.bridge_pubkey_bytes == bridge_pubkey_bytes && registration.sui_address != sender
 ) { 
  return false 
 }`

**Review by manolisliolios — APPROVED** [review]
LGTM

File: crates/sui-framework/packages/bridge/sources/committee.move (hunk 3 above)
Comment (dariorussi) [reply crates/sui-framework/packages/bridge/sources/committee.move:191]: not sure I follow, the found is to make sure that `bridge_pubkey_bytes` is unique in the list of proposed committee upgrade.
I may be missing something here

File: crates/sui-framework/packages/bridge/sources/committee.move (hunk 2 above)
Comment (dariorussi) [reply crates/sui-framework/packages/bridge/sources/committee.move:459]: I could, just leaving that in line with the other tests.
Having said so and given I need to review tests and code coverage I may apply that story to all existing tests in that future PR.
Thanks for the hint, I keep forgetting that and go trough tricks every time I write tests :)

File: crates/sui-framework/packages/bridge/sources/committee.move [hunk 4]
```diff
@@ -164,9 +165,35 @@ module bridge::committee {
             registration
         };
 
+        // check uniqeuness of the bridge pubkey.
```
Comment (longbowlu) [inline crates/sui-framework/packages/bridge/sources/committee.move:168]: uniqueness

File: crates/sui-framework/packages/bridge/sources/committee.move (hunk 1 above)
Comment (longbowlu) [reply crates/sui-framework/packages/bridge/sources/committee.move:172]: assert inside `check_uniqueness_bridge_keys` is slightly preferential to me, but not a strong opinion

File: crates/sui-framework/packages/bridge/sources/committee.move (hunk 3 above)
Comment (longbowlu) [reply crates/sui-framework/packages/bridge/sources/committee.move:191]: it messed with my brain as well and could be me malfunctioning too.

File: crates/sui-framework/packages/bridge/sources/committee.move (hunk 2 above)
Comment (longbowlu) [reply crates/sui-framework/packages/bridge/sources/committee.move:459]: til

**Review by longbowlu — APPROVED** [review]

File: crates/sui-framework/packages/bridge/sources/committee.move (hunk 4 above)
Comment (dariorussi) [reply crates/sui-framework/packages/bridge/sources/committee.move:168]: fixed

File: crates/sui-framework/packages/bridge/sources/committee.move (hunk 3 above)
Comment (dariorussi) [reply crates/sui-framework/packages/bridge/sources/committee.move:191]: renamed the field and added comments that should hopefully help fix the confusion

File: crates/sui-framework/packages/bridge/sources/committee.move (hunk 1 above)
Comment (dariorussi) [reply crates/sui-framework/packages/bridge/sources/committee.move]: changed

**Pull request merged**